From 89f62bba1957b93788cfd7ef7df5e869f0c741fd Mon Sep 17 00:00:00 2001
From: Yang Liu <numbksco@gmail.com>
Date: Mon, 28 Sep 2026 16:21:37 +0800
Subject: [PATCH 2/4] [DYNACACHE] Do not serialize when exiting from a fatal
 signal handler (#4482)

* [DYNACACHE] Do not serialize when exiting from a fatal signal handler

* review

(cherry picked from commit b80b3304c309e486ea5a1c3d26de34c15b4fff19)
---
 src/core.c                  | 1 +
 src/libtools/signal32.c     | 5 +++++
 src/libtools/signals.c      | 5 +++++
 src/os/emit_signals_linux.c | 4 ++++
 src/tools/dynacache.c       | 5 ++++-
 5 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/src/core.c b/src/core.c
index c33c384f..c84092f1 100644
--- a/src/core.c
+++ b/src/core.c
@@ -57,6 +57,7 @@ extern box64env_t box64env;
 
 int box64_quit = 0;
 int box64_exit_code = 0;
+int box64_in_fatal_handler = 0;
 int box64_stdout_no_w = 0;
 uintptr_t box64_pagesize;
 path_collection_t box64_addlibs = {0};
diff --git a/src/libtools/signal32.c b/src/libtools/signal32.c
index 35dfd046..ce55c4f7 100644
--- a/src/libtools/signal32.c
+++ b/src/libtools/signal32.c
@@ -58,6 +58,8 @@
 #endif
 
 
+extern int box64_in_fatal_handler;
+
 /* Definitions taken from the kernel headers.  */
 enum
 {
@@ -718,7 +720,10 @@ int my_sigactionhandler_oldcode_32(x64emu_t* emu, int32_t sig, int simple, sigin
     if(sig!=X64_SIGSEGV && !(Locks&is_dyndump_locked) && !(Locks&is_memprot_locked))
         dynarec = BOX64ENV(dynarec_interp_signal)?0:1;
     #endif
+    int in_fatal = (sig==X64_SIGSEGV) || (sig==X64_SIGBUS) || (sig==X64_SIGILL) || (sig==X64_SIGFPE);
+    if(in_fatal) ++box64_in_fatal_handler;
     ret = RunFunctionHandler32(&exits, dynarec, sigcontext, my_context->signals[info2->si_signo], 3, info2->si_signo, info2, sigcontext);
+    if(in_fatal && !exits && !emu->quit) --box64_in_fatal_handler;
     // restore old value from emu
     if(used_stack)  // release stack
         new_ss->ss_flags = 0;
diff --git a/src/libtools/signals.c b/src/libtools/signals.c
index 2745893b..8a3470db 100644
--- a/src/libtools/signals.c
+++ b/src/libtools/signals.c
@@ -47,6 +47,8 @@
 
 #include "signal_private.h"
 
+extern int box64_in_fatal_handler;
+
 static void sigstack_destroy(void* p)
 {
     x64_stack_t *ss = (x64_stack_t*)p;
@@ -549,7 +551,10 @@ int my_sigactionhandler_oldcode_64(x64emu_t* emu, int32_t sig, int simple, sigin
     if(!(/*sig==X64_SIGSEGV ||*/ (Locks&is_dyndump_locked) || (Locks&is_memprot_locked)))
         dynarec = BOX64ENV(dynarec_interp_signal)?0:1;
     #endif
+    int in_fatal = (sig==X64_SIGSEGV) || (sig==X64_SIGBUS) || (sig==X64_SIGILL) || (sig==X64_SIGFPE);
+    if(in_fatal) ++box64_in_fatal_handler;
     ret = RunFunctionHandler(emu, &exits, dynarec, sigcontext, my_context->signals[info2->si_signo], 3, info2->si_signo, info2, sigcontext);
+    if(in_fatal && !exits && !emu->quit) --box64_in_fatal_handler;
     if(used_stack)  // release stack
         new_ss->ss_flags = 0;
     // restore old value from emu
diff --git a/src/os/emit_signals_linux.c b/src/os/emit_signals_linux.c
index 3497dc18..fa33e9ca 100644
--- a/src/os/emit_signals_linux.c
+++ b/src/os/emit_signals_linux.c
@@ -25,6 +25,7 @@
 #include "box64cpu_util.h"
 
 void my_sigactionhandler_oldcode(x64emu_t* emu, int32_t sig, int simple, siginfo_t* info, void * ucntx, int* old_code, void* cur_db, uintptr_t x64pc);
+extern int box64_in_fatal_handler;
 void EmitSignal(x64emu_t* emu, int sig, void* addr, int code)
 {
     siginfo_t info = { 0 };
@@ -262,7 +263,10 @@ void EmitWineInt(x64emu_t* emu, int num, void* addr)
         sigcontext->uc_mcontext.gregs[X64_ERR] = 0x02|(num<<3);
         int exits = 0;
         int ret;
+        int in_fatal = (sig==X64_SIGSEGV) || (sig==X64_SIGBUS) || (sig==X64_SIGILL) || (sig==X64_SIGFPE);
+        if(in_fatal) ++box64_in_fatal_handler;
         ret = RunFunctionHandler(emu, &exits, 2, sigcontext, my_context->signals[info2->si_signo], 3, info2->si_signo, info2, sigcontext);
+        if(in_fatal && !exits && !emu->quit) --box64_in_fatal_handler;
         if(used_stack)  // release stack
             new_ss->ss_flags = 0;
         // restore values
diff --git a/src/tools/dynacache.c b/src/tools/dynacache.c
index 48160509..eebaf697 100644
--- a/src/tools/dynacache.c
+++ b/src/tools/dynacache.c
@@ -30,6 +30,8 @@
 #include "dynacache_hashes.h"
 #endif
 
+extern int box64_in_fatal_handler;
+
 #ifdef _WIN32
 #define PATHSEP "\\"
 #define HOME    "USERPROFILE"
@@ -91,7 +93,7 @@ done:
     `box64 --dynacache-clean` can be used from command line to purge obsolete DyaCache files
 */
 
-#define FILE_VERSION 6
+#define FILE_VERSION 7
 #define HEADER_SIGN  "DynaCache"
 
 typedef struct DynaCacheHeader_s {
@@ -434,6 +436,7 @@ void SerializeMmaplist(mapping_t* mapping)
     (void)mapping;
     return;
     #else
+    if(box64_in_fatal_handler) return;
     if(!DYNAREC_VERSION)
         return;
     if(mapping->env && mapping->env->is_dynacache_overridden && (mapping->env->dynacache!=1))
-- 
2.43.0

